### PR TITLE
Integrating new manual max invocation minters

### PR DESCRIPTION
**Pull request by lyaunzbe**
## Description of the change

Adding support for new max invocation minters.
Adding handler to index max invocation value from new minters.

>reminder: Any subgraph deployments should be documented as [Releases](https://github.com/ArtBlocks/artblocks-subgraph/releases) in this repository.

>reminder: Any changes to subgraph schema should be accompanied by corresponding changes to the Art Blocks documentation site, available at https://docs.artblocks.io/ (specifically, see the [Subgraph Entities](https://docs.artblocks.io/creator-docs/art-blocks-api/entities/) and [Subgraph Querying and API Overview](https://docs.artblocks.io/creator-docs/art-blocks-api/queries/) sections).

### CODE REVIEW

File: tests/subgraph/minter-suite/minter-suite-mapping.test.ts [hunk 1]
```diff
@@ -3174,3 +3178,168 @@ describe("MinterHolder-specific tests", () => {
...
+      )
+    ];
+
+    event.block.timestamp = CURRENT_BLOCK_TIMESTAMP;
+
+    handleProjectMaxInvocationsLimitUpdated(
+      changetype<ProjectMaxInvocationsLimitUpdated>(event)
+    );
+
+    const projectMinterConfigPostEvent = new ProjectMinterConfiguration(
```
Comment (yoshiwarab) [inline tests/subgraph/minter-suite/minter-suite-mapping.test.ts:3275]: This should be ProjectMinterConfiguration.load here

Comment (yoshiwarab) [reply tests/subgraph/minter-suite/minter-suite-mapping.test.ts:3275]: It doesn't look like you use it though?

File: tests/subgraph/minter-suite/minter-suite-mapping.test.ts [hunk 2]
```diff
@@ -3174,3 +3178,168 @@ describe("MinterHolder-specific tests", () => {
...
+      getProjectMinterConfigId(minterAddress.toHexString(), project.id)
+    );
+    assert.fieldEquals(
+      PROJECT_MINTER_CONFIGURATION_ENTITY_TYPE,
+      getProjectMinterConfigId(minterAddress.toHexString(), project.id),
```
Comment (yoshiwarab) [inline tests/subgraph/minter-suite/minter-suite-mapping.test.ts:3280]: Nit: should just save this in a variable so we don't need to regenerate in multiple places

**Review by yoshiwarab — APPROVED** [review]

File: tests/subgraph/minter-suite/minter-suite-mapping.test.ts (hunk 1 above)
Comment (lyaunzbe) [reply tests/subgraph/minter-suite/minter-suite-mapping.test.ts:3275]: ah yep, forgot to remove this (had a prev, incorrect test here)

File: src/minter-suite-mapping.ts [hunk 3]
```diff
@@ -153,6 +155,38 @@ export function handlePurchaseToDisabledUpdated(
...
+
+  if (!minterProjectAndConfig) {
+    return;
+  }
+
+  let project = minterProjectAndConfig.project;
+  let projectMinterConfig = minterProjectAndConfig.projectMinterConfiguration;
+
+  if (event.params._maxInvocations > project.maxInvocations) {
+    log.warning(
```
Comment (ryley-o) [inline src/minter-suite-mapping.ts]: optional - might be worth an in-line comment that says we don't expect minter contracts to allow this to happen

File: src/minter-suite-mapping.ts [hunk 4]
```diff
@@ -153,6 +155,38 @@ export function handlePurchaseToDisabledUpdated(
...
+  );
+
+  if (!minterProjectAndConfig) {
+    return;
+  }
+
+  let project = minterProjectAndConfig.project;
+  let projectMinterConfig = minterProjectAndConfig.projectMinterConfiguration;
+
+  if (event.params._maxInvocations > project.maxInvocations) {
```
[the remark this answers is not in the document]
Comment (lyaunzbe) [reply src/minter-suite-mapping.ts:175]: yes there's an implicit assumption that this cannot happen/the contract enforces this. will remove and just leave a comment instead

File: tests/subgraph/minter-suite/minter-suite-mapping.test.ts [hunk 5]
```diff
@@ -3174,3 +3178,168 @@ describe("MinterHolder-specific tests", () => {
...
+    );
+    projectMinterConfig.minter = minterAddress.toHexString();
+    projectMinterConfig.project = project.id;
+    projectMinterConfig.extraMinterDetails = "{}";
+    projectMinterConfig.basePrice = ONE_ETH_IN_WEI.div(BigInt.fromI32(10));
+    projectMinterConfig.priceIsConfigured = false;
+    projectMinterConfig.currencyAddress = Address.zero();
+    projectMinterConfig.currencySymbol = "ETH";
+    projectMinterConfig.purchaseToDisabled = false;
+    projectMinterConfig.maxInvocations = null;
```
Comment (ryley-o) [inline tests/subgraph/minter-suite/minter-suite-mapping.test.ts:3254]: nit: we don't actually set `projectMinterConfig.maxInvocations` to `null` in the `loadMinterProjectAndConfig` handler function, so shouldn't need to set to null here.

**Review by ryley-o — APPROVED** [review]

File: tests/subgraph/minter-suite/minter-suite-mapping.test.ts [hunk 6]
```diff
@@ -3174,3 +3178,168 @@ describe("MinterHolder-specific tests", () => {
...
+    event.block.timestamp = CURRENT_BLOCK_TIMESTAMP;
+
+    assert.notInStore(PROJECT_ENTITY_TYPE, fullProjectId);
+
+    handleProjectMaxInvocationsLimitUpdated(event);
+
+    assert.notInStore(PROJECT_MINTER_CONFIGURATION_ENTITY_TYPE, fullProjectId);
+  });
+
+  test("should do nothing if event maxInvocations are greater than core contract project maxInvocations", () => {
```
[the remark this answers is not in the document]
Comment (ryley-o) [reply tests/subgraph/minter-suite/minter-suite-mapping.test.ts]: (this is outdated after changes from Ben Y's review, so I'll resolve)

File: src/minter-suite-mapping.ts (hunk 3 above)
Comment (ryley-o) [reply src/minter-suite-mapping.ts]: this was addressed in Ben Y's review, so resolving

File: schema.graphql [hunk 7]
```diff
@@ -421,6 +422,9 @@ type ProjectMinterConfiguration @entity {
 
   "Configuration details used by specific minter project configurations (json string)"
   extraMinterDetails: String!
+
+  "Maximum number of invocations allowed for the project (on the minter)"
+  maxInvocations: BigInt
```
[the remark this answers is not in the document]
Comment (lyaunzbe) [reply schema.graphql:427]: yep, I think it couldn't hurt (and fwiw, I added a similar @dev comment already above the actual handler function

File: src/minter-suite-mapping.ts (hunk 3 above)
Comment (lyaunzbe) [reply src/minter-suite-mapping.ts]: based on the conversation above with @yoshiwarab , I actually did replace this entire conditional with a @dev comment

File: tests/subgraph/minter-suite/minter-suite-mapping.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (lyaunzbe) [reply tests/subgraph/minter-suite/minter-suite-mapping.test.ts]: ended up removing this test as unnecessary, but the comment is now present

File: tests/subgraph/minter-suite/minter-suite-mapping.test.ts (hunk 5 above)
Comment (lyaunzbe) [reply tests/subgraph/minter-suite/minter-suite-mapping.test.ts:3254]: 👍

**Pull request merged**